source/components/Options.js

```diff
@@ -0,0 +1,264 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import ReactDOM from 'react-dom';
+import SkinnableComponent from './SkinnableComponent';
+import events from '../utils/events';
+
+export default class Options extends SkinnableComponent {
+
+  static SKIN_PARTS = {
+    OPTIONS: 'options',
+  };
+
+  static propTypes = Object.assign({}, SkinnableComponent.propTypes, {
+    isOpen: PropTypes.bool,
+    isOpeningUpward: PropTypes.bool,
+    resetOnClose: PropTypes.bool, // reset highlighted option on options close (e.g. in autocomplete)
+    optionRenderer: PropTypes.func,
+    selectedOptionValue: PropTypes.string,
+    noResults: PropTypes.bool,
+  });
+
+  static defaultProps = {
+    isOpen: false,
+    isOpeningUpward: false,
+    resetOnClose: false,
+  };
+
+  state = {
+    isOpen: this.props.isOpen,
+    highlightedOptionIndex: 0,
+    position: null,
+  };
+
+  componentWillReceiveProps (nextProps) {
+    if (this.props.isOpen !== nextProps.isOpen ) {
+      this.updateComponentsStates(nextProps.isOpen, null);
+    }
+  }
+
+  componentWillUpdate (nextProps, nextState) {
+    // update isOpen state when parent component force open / close options (e.g. click on Input in Select component)
+    if (!this.state.isOpen && nextState.isOpen) {
+      window.addEventListener("resize", this._handleWindowResize);
+      this.handleScrollEventListener('add');
+      events.addEventsToDocument(this._getDocumentEvents());
+    }
+  }
+
+  componentDidUpdate (prevProps, prevState) {
+    if (prevState.isOpen && !this.state.isOpen) {
+      this.handleScrollEventListener('remove');
+      events.removeEventsFromDocument(this._getDocumentEvents());
+      window.removeEventListener("resize", this._handleWindowResize);
+    }
+
+    if (this.props.isOpen !== prevProps.isOpen ) {
+      const rootNode = this._getRootSkinPart();
+      const parentNode = this._getRootSkinPart().parentNode;
+      const rootNodeParams = rootNode.getBoundingClientRect();
+      const parentNodeParams = parentNode.getBoundingClientRect();
+
+      let positionY;
+      if (this.props.isOpeningUpward) {
+        positionY = parentNodeParams.bottom - parentNodeParams.height - rootNodeParams.height + 20;
+      } else {
+        positionY = parentNodeParams.bottom + 20;
+      }
+
+      const position = {
+        width: parentNodeParams.width,
+        positionX: parentNodeParams.left,
+        positionY,
+      };
+      this.updateComponentsStates(this.props.isOpen, position);
+    }
+  }
+
+  componentWillUnmount () {
+    if (this.state.isOpen) {
+      events.removeEventsFromDocument(this._getDocumentEvents());
+      window.removeEventListener("resize", this._handleWindowResize);
+      this.handleScrollEventListener('remove');
```

Comment:
These 3 lines should be moved into a separate method `removeAllEventListeners` because it is duplicated (a few lines above in `componentDidUpdate`) 😉 

```javascript
removeAllEventListeners () {
  events.removeEventsFromDocument(this._getDocumentEvents());
  window.removeEventListener("resize", this._handleWindowResize);
  this.handleScrollEventListener('remove');
}
```

Reply:
ahaa I see, will fix that ;)